src/grpc/header_interceptor.rs

```diff
@@ -0,0 +1,35 @@
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicBool, Ordering};
+
+const AUTHORIZATION: &str = "authorization";
+const AGENT: &str = "agent";
+
+#[derive(Clone)]
+pub struct HeaderInterceptor {
+    pub header: HashMap<String, String>,
+}
+
+impl tonic::service::Interceptor for HeaderInterceptor {
+    fn call(
+        &mut self,
+        mut request: tonic::Request<()>,
+    ) -> Result<tonic::Request<()>, tonic::Status> {
+        static ARE_ONLY_ONCE_HEADER_SENT: AtomicBool = AtomicBool::new(false);
+        for (key, value) in self.header.iter() {
+            if *key == *AUTHORIZATION {
+                request.metadata_mut().insert(
+                    tonic::metadata::AsciiMetadataKey::from_static(AUTHORIZATION),
+                    tonic::metadata::AsciiMetadataValue::from_str(value).unwrap(),
+                );
+            }
+            if !ARE_ONLY_ONCE_HEADER_SENT.load(Ordering::Relaxed) && *key != *AUTHORIZATION {
+                request.metadata_mut().insert(
+                    tonic::metadata::AsciiMetadataKey::from_static(AGENT),
```

Comment:
Should this be? 
`tonic::metadata::AsciiMetadataKey::from_static(*key)`

What if we add another header to `ARE_ONLY_ONCE_HEADER_SENT` other then agent down line?

Reply:
Since `tonic::metadata::AsciiMetadataKey::from_static` takes only a static variable, `AGENT` needs to get passed, I believe. (At least passing `*key` is not valid)
Also `AGENT` is a constant variable which lives the entire lifetime of a program, so it can be used later like the below example:
```
if !ARE_ONLY_ONCE_HEADER_SENT.load(Ordering::Relaxed) && *key != *AUTHORIZATION {
     request.metadata_mut().insert(
     tonic::metadata::AsciiMetadataKey::from_static(AGENT),
     tonic::metadata::AsciiMetadataValue::from_str(value).unwrap(),
      );
      ARE_ONLY_ONCE_HEADER_SENT.store(true, Ordering::Relaxed);
}
request.metadata_mut().insert(
     tonic::metadata::AsciiMetadataKey::from_static(AGENT),
     tonic::metadata::AsciiMetadataValue::from_str(value).unwrap(),
);
```

Reply:
shouldn't you switch up logic in if statement then to be like
`if !ARE_ONLY_ONCE_HEADER_SENT.load(Ordering::Relaxed) && *key == *AGENT {`

Reply:
Yup, it makes sense. 
There might be a better way to handle new metadata in the future but I spent some time but this `&mut self` thing is a bit confusing and needing to pass a static variable was the part I just time boxed yesterday.
I think it's okay for now to specifically call `AGENT` we can come back and update code structure later if needed.